File: trunk/drools/drools-core/src/main/org/drools/RuleBaseBuilder.java
package org.drools;

import org.drools.rule.Rule;
import org.drools.rule.RuleSet;
import org.drools.reteoo.Builder;
import org.drools.reteoo.Rete;

public class RuleBaseBuilder
{
    private Rete rete;
    private Builder builder;

    public RuleBaseBuilder()
    {
    }

    public void addRule(Rule rule)
        throws RuleIntegrationException
    {
        init();
        this.builder.addRule( rule );
    }

    public void addRuleSet(RuleSet ruleSet)
        throws RuleIntegrationException
    {
        init();
        
        Rule[] rules = ruleSet.getRules();

        for ( int i = 0 ; i < rules.length ; ++i )
        {
            this.builder.addRule( rules[ i ] );
        }
    }

    public RuleBase build()
    {
        Rete rete= this.rete;

        this.rete    = null;
        this.builder = null;

        return new RuleBaseImpl( this.rete );
    }

    void init()
    {
        if ( this.rete == null )
        {
            this.rete    = new Rete();
            this.builder = new Builder( this.rete );
        }
    }
}

File: trunk/drools/drools-dotnet/src/drools.dotnet/RuleBaseBuilder.cs
using System;
using org.drools.dotnet.rule;
#if FRAMEWORK11
	using System.Collections;
#else
	using System.Collections.Generic;
#endif

namespace org.drools.dotnet
{
	public class RuleBaseBuilder
	{
#if FRAMEWORK11
		private IList _ruleSets = new ArrayList();
#else
		private ICollection<RuleSet> _ruleSets = new List<RuleSet>();
#endif
		public RuleBaseBuilder()
		{
		}

#if FRAMEWORK11
		public IList RuleSets
#else
		public ICollection<RuleSet> RuleSets
#endif
		{
			get { return _ruleSets; }
		}

		public RuleBase Build()
		{
			drools.RuleBaseBuilder rbb = new org.drools.RuleBaseBuilder();
			foreach (RuleSet ruleSet in _ruleSets)
			{
				rbb.addRuleSet(ruleSet.InternalRuleSet);
			}
			return new RuleBase(rbb.build());
		}
	}
}
